libs/db/boltdb.go

```diff
@@ -150,36 +149,58 @@ func (bdb *BoltDB) Stats() map[string]string {
 // boltDBBatch stores key values in sync.Map and dumps them to the underlying
 // DB upon Write call.
 type boltDBBatch struct {
-	buffer *sync.Map
-	db     *BoltDB
+	buffer []struct {
+		k []byte
+		v []byte
+	}
+	db *BoltDB
 }
 
 // NewBatch returns a new batch.
 func (bdb *BoltDB) NewBatch() Batch {
 	return &boltDBBatch{
-		buffer: &sync.Map{},
-		db:     bdb,
+		buffer: make([]struct {
+			k []byte
+			v []byte
+		}, 0),
+		db: bdb,
 	}
 }
 
+// It is safe to modify the contents of the argument after Set returns but not
+// before.
 func (bdb *boltDBBatch) Set(key, value []byte) {
-	bdb.buffer.Store(string(key), value)
+	bdb.buffer = append(bdb.buffer, struct {
+		k []byte
+		v []byte
```

Comment:
I guess you didn't want to use `KVPair` because it is protobuf and you do not need any encoding here etc?

Reply:
yes